GitUpKit/Views/GIAdvancedCommitViewController.m

```diff
@@ -255,18 +264,31 @@ - (void)_diffFilesViewControllerDidPressDelete:(GIDiffFilesViewController*)contr
                                     button:NSLocalizedString(@"Discard", nil)
                  suppressionUserDefaultKey:nil
                                      block:^{
+                                       NSMutableArray* selectedFiles = [NSMutableArray array];
+
                                        for (GCDiffDelta* delta in deltas) {
                                          NSError* error;
                                          BOOL submodule = delta.submodule;
-                                         if ((submodule && ![self discardSubmoduleAtPath:delta.canonicalPath resetIndex:NO error:&error]) || (!submodule && ![self discardAllChangesForFile:delta.canonicalPath resetIndex:NO error:&error])) {
-                                           [self presentError:error];
-                                           break;
+                                         if (submodule) {
+                                           // We handle every submodules deltas individually
+                                           if (![self discardSubmoduleAtPath:delta.canonicalPath resetIndex:NO error:&error]) {
+                                             [self presentError:error];
+                                             break;
+                                           }
+                                         } else {
+                                           // Otherwise we collect file delta paths to batch process them afterwards
+                                           [selectedFiles addObject:delta.canonicalPath];
                                          }
                                        }
-                                       [self.repository notifyWorkingDirectoryChanged];
-                                       if (!_workdirFilesViewController.deltas.count) {
-                                         _indexActive = YES;
-                                         [self.view.window makeFirstResponder:_indexFilesViewController.preferredFirstResponder];
+
+                                       NSError* error;
+                                       if (![self discardAllChangesForFiles:selectedFiles resetIndex:NO error:&error]) {
+                                         [self.repository notifyWorkingDirectoryChanged];
```

Comment:
I believe this was the issue around why it didn't notify when discarding files. This code would only notify on failures basically. The other issue where it would discard things I didn't select was the most concerning however.

Reply:
Ugh yeah the `notifyWorkingDirectoryChanged` and `makeFirstResponder` lines definitely shouldn't be in that `if`.

I'll review the rest of the code and see if I can reproduce that otgher issue where more things were discarded than necessary.

Reply:
I really can't reproduce the other issue (with these lines placed at the right place). I wonder if it might be related to your other issue where the wrong files were discarded though. I'll revert my fix and see if I can make it happen like this.

Reply:
Okaaaaayyyy I've reproduced it. It's bad.

Reply:
Looks like the issue is in the call to the libgit2 function `git_checkout_index`. Right before it's called, the repo is fine (the right files have been deleted) but it seems like it resets too many files.